enumerator.c

```diff
@@ -2028,6 +2031,75 @@ lazy_map(VALUE obj)
     return lazy_add_method(obj, 0, 0, Qnil, Qnil, &lazy_map_funcs);
 }
 
+static struct MEMO *
+lazy_scan_left_proc(VALUE proc_entry, struct MEMO *result, VALUE memos, long memo_index)
+{
+    struct proc_entry *entry = proc_entry_ptr(proc_entry);
+    VALUE memo = rb_ary_entry(memos, memo_index);
+    VALUE argv[2];
+    VALUE value;
+
+    if (NIL_P(memo)) {
+        memo = result->memo_value;
+        rb_ary_store(memos, memo_index, memo);
+
+        return result;
+    }
+
+    argv[0] = memo;
+    argv[1] = result->memo_value;
+
+    value = rb_proc_call_with_block(entry->proc, 2, argv, Qnil);
+    LAZY_MEMO_SET_VALUE(result, value);
+    rb_ary_store(memos, memo_index, value);
+    return result;
+}
+
+static const lazyenum_funcs lazy_scan_left_funcs = {
+    lazy_scan_left_proc, 0,
+};
+
+/*
+ *  call-seq:
+ *     lazy.scan_left(initial) { |memo, obj| block } -> lazy_enumerator
+ *     lazy.scan_left          { |memo, obj| block } -> lazy_enumerator
+ *
+ *  Like Enumerable#scan_left, but chains operation to be lazy-evaluated.
+ *
+ *     (1..Float::INFINITY).lazy.scan_left(0) { |sum, i| sum + i }
+ *     #=> #<Enumerator::Lazy: #<Enumerator::Lazy: #<Enumerator::Chain: [[0], #<Enumerator::Lazy: 1..Infinity>]>>:scan_left>
+ *     (1..Float::INFINITY).lazy.scan_left(0) { |sum, i| sum + i }.first(3)
+ *     #=> [0, 1, 3]
+ *
+ *     #=> #<Enumerator::Lazy: #<Enumerator::Lazy: 1..Infinity>:scan_left>
+ *     (1..Float::INFINITY).lazy.scan_left { |sum, i| sum + i }
+ *     #=> #<Enumerator::Lazy: #<Enumerator::Lazy: 1..Infinity>:scan_left>
+ *     (1..Float::INFINITY).lazy.scan_left { |sum, i| sum + i }.first(3)
+ *     #=> [1, 3, 6]
+ */
+
+static VALUE
+lazy_scan_left(int argc, VALUE *argv, VALUE obj)
+{
+    VALUE memo, extended_enum, enums;
+
+    if (!rb_block_given_p()) {
+	rb_raise(rb_eArgError, "tried to call lazy scan left without a block");
+    }
+
+    rb_scan_args(argc, argv, "01", &memo);
+
+    if (NIL_P(memo)) {
+        return lazy_add_method(obj, 0, 0, Qnil, Qnil, &lazy_scan_left_funcs);
+    }
+
+    enums = rb_ary_new3(1, rb_ary_new3(1, memo));
+    rb_ary_push(enums, obj);
+    extended_enum = enumerable_lazy(enum_chain_initialize(enum_chain_allocate(rb_cEnumChain), enums));
```

Comment:
I don't know the best way to get lazy behavior while prepending the initial argument. This is a hacky approach that wraps the argument in an array and then chains that array with the lazy enumerable, then scans over the resulting chained enumerable.

I am very happy to receive any suggestions on how to simplify or improve this!